bubuku/features/rebalance/change.py

```diff
@@ -0,0 +1,248 @@
+            if target.accept_replica(broker, topic_partition):
+                self.action_queue.append((topic_partition, broker.broker_id, target.broker_id))
+                return True
+        return False
+
+    def _rebalance_replicas_template(self, force: bool):
+        for broker in self.broker_distribution.values():
+            to_move = broker.get_replica_overload()
+            if to_move <= 0:
+                continue
+            targets = self._list_active_brokers_with_skip(broker.broker_id)
+            if not force:
+                targets = [t for t in targets if t.has_free_replica_slots()]
+            for _ in range(0, to_move):
+                copied = False
+                for topic_partition in broker.list_replicas():
+                    if self._move_partition_to_replica(broker, topic_partition, targets):
+                        copied = True
+                        break
+                if not copied:
+                    return False
+                if not force:
+                    targets = [t for t in targets if t.has_free_replica_slots()]
+        return True
+
+    @staticmethod
+    def _build_leader_distribution_weights(brokers: dict):
+        candidates = {}
+        for source_broker_id, source_broker in brokers.items():
+            if not source_broker.have_extra_leaders():
+                continue
+            for target_broker_id, target_broker in brokers.items():
+                if not target_broker.have_less_leaders():
+                    continue
+                weight_list = []
+                for topic in source_broker.topic_cardinality.keys():
+                    delta = source_broker.topic_cardinality[topic] - target_broker.topic_cardinality.get(topic, 0)
+                    weight_list.append((topic, delta))
+                # Now the first element is the first to rebalance.
+                if weight_list:
+                    candidates[(source_broker_id, target_broker_id)] = sorted(
+                        weight_list, key=lambda x: x[1], reverse=True)
+        return candidates
+
+    @staticmethod
+    def _place_to_cardinalities(array: list, topic: str, cardinality_change: int):
+        old_cardinality = None
+        for item in [(a, b) for a, b in array if a == topic]:
+            old_cardinality = item[1]
+            array.remove(item)
+        if old_cardinality is None:
+            return
+        new_cardinality = old_cardinality + cardinality_change
+        idx = len(array)
+        for i in range(idx - 1, -1, -1):
+            if array[i][1] >= new_cardinality:
+                idx = i + 1
+                break
+        array.insert(idx, (topic, new_cardinality))
+
+    @staticmethod
+    def rebalance_leaders(brokers: dict) -> list:
+        candidates = OptimizedRebalanceChange._build_leader_distribution_weights(brokers)
+        queue_list = []
+        while any([broker.have_extra_leaders() for broker in brokers.values()]):
```

Comment:
Is it guaranteed that extra leader is always decremented in this loop? It's a very long loop body and it was hard for me to verify that. If there is any problem, we would have an infinite loop. I would be more comfortable if we broke it down into smaller steps easier to verify.